docs/model_developers/references.rst

```diff
@@ -0,0 +1,132 @@
+.. _references:
+
+References and Secrets
+######################
+
+When a handler needs information that is not available to the compiler, references are used.
+
+For example, to extract username and password from an environment variables:
+
+.. code-block:: inmanta
+    :emphasize-lines: 6-7
+
+    leaf1 = nokia_srlinux::GnmiDevice(
+        auto_agent=true,
+        name="leaf1",
+        mgmt_ip="172.30.0.210",
+        yang_credentials=yang::Credentials(
+            username=std::create_environment_reference("GNMI_USER"),
+            password=std::create_environment_reference("GNMI_PASS"),
+        ),
+    )
+
+
+This means that the username and password will never be present in the compiler, but that they will be picked up by the handler when needed.
+This is very different from using ``std::get_env``, which will resolve the environment variable in the compiler and store it in the database.
+
+More advanced combination are also possible:
+
+.. code-block:: inmanta
+    :emphasize-lines: 1,3,13-14
+
+    netbox_secret = netbox::create_netbox_reference(
+        netbox_url=...,
+        netbox_token=std::create_environment_reference("NETBOX_API_TOKEN"),
+        device=leaf1.name,
+        role="admin",
+    )
+
+    leaf1 = nokia_srlinux::GnmiDevice(
+        auto_agent=true,
+        name="leaf1",
+        mgmt_ip="172.30.0.210",
+        yang_credentials=yang::Credentials(
+            username=netbox_secret.name,
+            password=netbox_secret.password,
+        ),
+    )
+
+Here we get the root secret from the environment variable ``NETBOX_API_TOKEN``, which we use to query the inventory for proper credentials for this device.
+
+References can be used in the model and in resources transparently.
+However any attempt to perform an operation (e.g. addition, string formatting,... ) on a reference will result in an exception.
+
+References will be automatically resolved before passing the resource into a handler.
+I.e. using references in handlers requires no special attention.
+
+
+Creating new types of References
+---------------------------------
+
+When you want to expose your own type of reference, the following steps are required:
+
+1. Create a subclass of :py:class:`inmanta.references.Reference`, with a type parameter that is either a primitive type or a dataclass.
+2. Annotate it with :py:func:`inmanta.references.reference`.
+3. Implement the ``resolve`` method, that will resolve the reference. This method can use ``self.resolve_others`` to resolve any reference received as an argument. The logger passed into this method is similar to the ``ctx`` argument passed into a handler. The logger will write the logs into the database as well, when resolving references for a handler.
+4. Create a plugin to construct the reference.
```

Comment:
Did fixing the numbering fix the way it is rendered ?

Reply:
it did